lang/cs/Org.Apache.REEF.Network/Utilities/GenericInMemoryCache.cs

```
@@ -0,0 +1,355 @@
+
+            locker.EnterReadLock();
+            try
+            {
+                T rv = cache.TryGetValue(key, out rv) ? rv : default(T);
+                return rv;
+            }
+            finally
+            {
+                locker.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        /// Tries to gets the cache entry with the specified key.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <param name="value">(out) The value, if found, or <c>default(T)</c>, if not.</param>
+        /// <returns><c>True</c>, if <c>key</c> exists, otherwise <c>false</c>.</returns>
+        public bool TryGet(K key, out T value)
+        {
+            if (disposed)
+            {
+                value = default(T);
+                return false;
+            }
+
+            locker.EnterReadLock();
+            try
+            {
+                return cache.TryGetValue(key, out value);
+            }
+            finally
+            {
+                locker.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        /// Removes a series of cache entries in a single call for all key that match the specified key pattern.
+        /// </summary>
+        /// <param name="keyPattern">The key pattern to remove. The Predicate has to return true to get key removed.</param>
+        public void Remove(Predicate<K> keyPattern)
+        {
+            if (disposed)
+            {
+                return;
+            }
+
+            locker.EnterWriteLock();
+            try
+            {
+                var removers = (from k in cache.Keys.Cast<K>()
+                                where keyPattern(k)
+                                select k).ToList();
+
+                foreach (K workKey in removers)
+                {
+                    try
+                    {
+                        timers[workKey].Dispose();
+                    }
+                    catch
+                    {
```

ignore the exception?